homeassistant/components/here_travel_time/sensor.py

```diff
@@ -0,0 +1,285 @@
+
+        # pylint: disable=E1101
+        _summary = self._response.response['route'][0]['summary']
+        return round(_summary['trafficTime'] / 60)
+
+    @property
+    def name(self):
+        """Get the name of the sensor."""
+        return self._name
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        if self._response is None:
+            return None
+
+        # pylint: disable=E1101
+        _summary = self._response.response['route'][0]['summary']
+        # pylint: disable=E1101
+        _route = self._response.response['route']
+
+        res = {}
+        res['distance'] = _summary['distance']
+        res['distance_unit'] = LENGTH_METERS
+        res['trafficTime'] = _summary['trafficTime']
+        res['baseTime'] = _summary['baseTime']
+        res['travelTime'] = _summary['travelTime']
+        res['origin_name'] = _route[0]['waypoint'][0]['mappedRoadName']
+        res['destination_name'] = _route[0]['waypoint'][1]['mappedRoadName']
+        res[CONF_MODE] = self._travel_mode
+        res[CONF_TRAFFIC_MODE] = self._traffic_mode
+        return res
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit this state is expressed in."""
+        return self._unit_of_measurement
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):
+        """Get the latest data from HERE."""
+        import herepy
+        # Convert device_trackers to HERE friendly location
+        if hasattr(self, '_origin_entity_id'):
+            self._origin = self._get_location_from_entity(
+                self._origin_entity_id
+            )
+
+        if hasattr(self, '_destination_entity_id'):
+            self._destination = self._get_location_from_entity(
+                self._destination_entity_id
+            )
+
+        self._destination = self._resolve_zone(self._destination)
+        self._origin = self._resolve_zone(self._origin)
+
+        # Convert location to HERE friendly location if not already so
+        if not isinstance(self._destination, list):
+            self._destination = self._destination.split(',')
+        if not isinstance(self._origin, list):
+            self._origin = self._origin.split(',')
+
+        if self._traffic_mode:
+            traffic_mode = TRAFFIC_MODE_ENABLED
+        else:
+            traffic_mode = TRAFFIC_MODE_DISABLED
+
+        if self._destination is not None and self._origin is not None:
+            response = self._client.car_route(
+                self._origin,
+                self._destination,
+                [self._travel_mode, self._route_mode, traffic_mode],
+            )
+            if isinstance(response, herepy.error.HEREError):
+                _LOGGER.error("API returned error %s", response.message)
+                self.valid_api_connection = False
+                return
+
+            self._response = response
+
+    def _get_location_from_entity(self, entity_id):
```

Comment:
samer here about generalized travel_time component isn't this a ton of dup code?